src/megaclient.cpp

```diff
@@ -13705,19 +14081,269 @@ void MegaClient::execmovetosyncdebris()
+{
+    if (syncConfigs)
+    {
+        assert(config);
+        auto newConfig = *config;
+
+        auto isEnabled = [](syncstate_t state, syncerror_t syncError) -> bool
+        {
+            return state != SYNC_CANCELED && (state != SYNC_DISABLED || syncError != NO_SYNC_ERROR );
+        };
+
+        newConfig.setEnabled(isEnabled(newstate, newSyncError));
+        newConfig.setError(newSyncError);
+
+        syncConfigs->insert(newConfig);
+        return API_OK;
+    }
+    return API_ENOENT;
+}
+
+error MegaClient::changeSyncState(const SyncConfig *config, syncstate_t newstate, syncerror_t newSyncError, bool fireDisableEvent)
+{
+    error e = API_OK;
+    assert(config);
+    if (config)
+    {
+        auto isEnabled = [](syncstate_t state, syncerror_t syncError) -> bool
+        {
+            return state != SYNC_CANCELED && (state != SYNC_DISABLED || syncError != NO_SYNC_ERROR );
+        };
+
+        if ( (config->getError() != newSyncError) || (config->getEnabled() != isEnabled(newstate, newSyncError)) ) //has changed
+        {
+            e = saveAndUpdateSyncConfig(config, newstate, newSyncError);
+        }
+    }
+    else
+    {
+        e = API_ENOENT;
+    }
+
+    app->syncupdate_state(config->getTag(), newstate, newSyncError, fireDisableEvent);
+    return e;
+}
+
+error MegaClient::changeSyncState(int tag, syncstate_t newstate, syncerror_t newSyncError, bool fireDisableEvent)
+{
+    error e = API_OK;
+
+    if (!syncConfigs)
+    {
+        LOG_err << "no SyncConfig upon changeSyncState";
+        e = API_ENOENT;
+    }
+    auto config = syncConfigs?syncConfigs->get(tag):nullptr;
+    assert(config);
+    e = changeSyncState(config, newstate, newSyncError, fireDisableEvent);
+    return e;
+}
+
+error MegaClient::changeSyncStateByNodeHandle(mega::handle nodeHandle, syncstate_t newstate, syncerror_t newSyncError, bool fireDisableEvent)
+{
+    error e = API_OK;
+
+    auto tag = 0;
+
+    if (!syncConfigs)
+    {
+        LOG_err << "no SyncConfig upon changeSyncStateByNodeHandle";
+        e = API_ENOENT;
+    }
+
+    auto config = syncConfigs?syncConfigs->getByNodeHandle(nodeHandle):nullptr;
+    assert(config);
+    e = changeSyncState(config, newstate, newSyncError, fireDisableEvent);
+
+    return API_ENOENT;
+}
+
+void MegaClient::failSync(Sync* sync, syncerror_t syncerror)
+{
+    LOG_err << "Failing sync: " << sync->getConfig().getLocalPath() << " error = " << syncerror;
+
+    sync->changestate(SYNC_FAILED, syncerror); //This will cause the later deletion of Sync (not MegaSyncPrivate) object
+
+    syncactivity = true;
+}
+
+void MegaClient::disableSync(Sync* sync, syncerror_t syncError)
+{
+    sync->errorCode = syncError;
+    sync->changestate(SYNC_DISABLED, syncError); //This will cause the later deletion of Sync (not MegaSyncPrivate) object
+
+    syncactivity = true;
+}
+
+bool MegaClient::disableSyncContainingNode(mega::handle nodeHandle, syncerror_t syncError)
```

Comment:
Maybe it doesn't worth, but a method to find a sync based on a node could be handy in cases like this one. It would result on a code that is like:
``` cpp
    Sync *sync = getSyncByNodehandle(handle);
    if (sync)
        disableSync(sync, syncError);
```